### PR TITLE
`AbstractItem#doReload` should not be calling `SaveableListener#fireOnChange`

### PR DESCRIPTION
**Pull request by Vlatombe**
Was initially added as ce7e6390ee29ab59ebff92a8f3033c9e5ef02f4e

`SaveableListener.fireOnChange` is typically called when saving an object, not when loading it.
This call should be removed as it causes unnecessary calls to listeners when in fact no change happened.

### Proposed changelog entries

- Developer: do not call `SaveableListener.fireOnChange` anymore when reloading an `AbstractItem`.

### Proposed upgrade guidelines

Upgrade notes: Jenkins config history plugin users: when reloading an individual job from CLI, the event is no longer recorded by the plugin. Note that it was already the case when reloading all jobs through the "Reload configuration from disk" action.

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [X] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7984"></a>

### CODE REVIEW

**Review by jglick — COMMENTED** [review]
Looks right. Can it be verified around https://github.com/jenkinsci/jenkins/blob/cd3b1517e7a553fd6fcf39e25c4bd1e18fbe39de/test/src/test/java/hudson/model/AbstractItemTest.java#L54?

File: test/src/test/java/hudson/model/AbstractItemTest.java [hunk 1]
```diff
@@ -143,4 +154,25 @@ private String getPath(URL u) {
         return u.getPath().substring(j.contextPath.length() + 1);
     }
 
+    @TestExtension
```
Comment (jglick) [inline test/src/test/java/hudson/model/AbstractItemTest.java:157]:
```suggestion
    @TestExtension("reload')
```

**Review by jglick — APPROVED** [review]

**Comment by daniel-beck** [comment]
Something like https://plugins.jenkins.io/jobConfigHistory/ probably relies on this event to register a change to a configuration file from disk. Are we dropping support (however implicit it was) for externally modifying files, followed by a reload? Or how would that be approached going forward?

**Comment by Vlatombe** [comment]
> Are we dropping support (however implicit it was) for externally modifying files, followed by a reload?

Let me check that scenario

**Comment by jglick** [comment]
If that is actually needed, I think it could be supported with a new `SaveableListener.onReload(Saveable, XmlFile)`.

**Comment by Vlatombe** [comment]
> Let me check that scenario

It doesn't seem to record any new revision after reloading an external change (with current core#master). However it detects the condition and prints a warning "This is the latest revision in history, but not the current config. Save config to create a current entry."

**Comment by Vlatombe** [comment]
> Let me check that scenario

It doesn't seem to record any new revision after reloading an external change (with current core#master). However it detects the condition and prints a warning "This is the latest revision in history, but not the current config. Save config to create a current entry."

Double checking, actually the whole reload of Jenkins does NOT call this listener, while reloading a single item does. Even if jobConfigHistory could leverage this when reloading a single item, the behaviour is inconsistent.

**Comment by daniel-beck** [comment]
@Vlatombe Could you clarify whether this is causing a regression with JCH or not? It seems it is, specifically for reloading individual jobs?

(Not blocking, but decision to merge should be made while aware of this.)

**Comment by Vlatombe** [comment]
> It seems it is, specifically for reloading individual jobs?

Yes, this is the only impacted action, though it is not exposed through the UI, only via CLI.

**Comment by jglick** [comment]
> to do things well it would also require to call this new listener when…

Right, does not sound straightforward.

I doubt that recording changes _on disk_ from JCH is an important use case. The plugin’s sweet spot is keeping track of haphazard configuration edits coming from the web UI, especially multiple users. If you are managing configuration on disk it is likely because you are using some as-code system (but not `configuration-as-code`, which calls setters that `save`) which is either backed by SCM already or should be. Would suffice to mention the change in upgrade notes IMO.

**Review by timja — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Pull request merged**